File: app/services/workspaces.py
from __future__ import annotations

import re

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.config import settings
from app.models import User, Workspace


def slugify(value: str) -> str:
    slug = re.sub(r"[^a-z0-9]+", "-", value.lower()).strip("-")
    return slug or "workspace"


def create_workspace(db: Session, user: User, name: str) -> Workspace:
    base_slug = slugify(name)
    slug = base_slug
    counter = 2
    while db.scalar(select(Workspace).where(Workspace.slug == slug)) is not None:
        slug = f"{base_slug}-{counter}"
        counter += 1

    workspace = Workspace(
        user_id=user.id,
        name=name.strip()[:120] or "Moj workspace",
        slug=slug,
        daily_quota=settings.default_daily_quota,
    )
    db.add(workspace)
    db.commit()
    db.refresh(workspace)
    return workspace


def create_default_workspace(db: Session, user: User) -> Workspace:
    label = (user.name or user.email.split("@")[0]).strip()
    return create_workspace(db, user, f"Workspace {label}"[:120])


def list_workspaces(db: Session, user: User) -> list[Workspace]:
    return list(
        db.scalars(select(Workspace).where(Workspace.user_id == user.id).order_by(Workspace.id))
    )

File: src/services/workspaces.ts
import { config } from "../config.js";
import { db } from "../db/index.js";
import type { User, Workspace } from "../db/types.js";

export function slugify(value: string): string {
  const slug = value
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");
  return slug || "workspace";
}

export async function createWorkspace(userId: number, name: string): Promise<Workspace> {
  const baseSlug = slugify(name);
  let slug = baseSlug;
  let counter = 2;
  // Slug trafia do adresow URL, wiec musi byc unikalny w calej instalacji.
  while (
    await db.selectFrom("workspaces").select("id").where("slug", "=", slug).executeTakeFirst()
  ) {
    slug = `${baseSlug}-${counter}`;
    counter += 1;
  }

  const result = await db
    .insertInto("workspaces")
    .values({
      user_id: userId,
      name: name.trim().slice(0, 120) || "Moj workspace",
      slug,
      daily_quota: config.defaultDailyQuota,
    })
    .executeTakeFirst();

  const workspace = await db
    .selectFrom("workspaces")
    .selectAll()
    .where("id", "=", Number(result.insertId))
    .executeTakeFirstOrThrow();
  return workspace;
}

export function createDefaultWorkspace(user: Pick<User, "id" | "name" | "email">): Promise<Workspace> {
  const label = (user.name || user.email.split("@")[0] || "").trim();
  return createWorkspace(user.id, `Workspace ${label}`.slice(0, 120));
}

export function listWorkspaces(userId: number): Promise<Workspace[]> {
  return db
    .selectFrom("workspaces")
    .selectAll()
    .where("user_id", "=", userId)
    .orderBy("id")
    .execute();
}

export function getWorkspace(workspaceId: number, userId: number): Promise<Workspace | undefined> {
  return db
    .selectFrom("workspaces")
    .selectAll()
    .where("id", "=", workspaceId)
    .where("user_id", "=", userId)
    .executeTakeFirst();
}
